### PR TITLE
feature: Add Docker build

### LINKED ISSUE #51: Create a Docker file to run the backend
## Description
As a developer,
I need a way to run the backend using Docker,
so that I can run the project with one command without the need to setup my local dev environment.

### Update [Required]
- [ ] Create a Dockerfile to run the project using one command
- [ ] Add documentation on running the program using Docker on README

## Definition of Done
- [ ] All of the required items are completed.
- [x] Approval by 1 mentor.

## Estimation
1 hours

### PR DESCRIPTION
**Pull request by devkapilbansal**
### Description

Fixes #51

### Type of Change:

- Quality Assurance
- Documentation

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by isabelcosta** [comment]
I know its just a draft, but this is so exciting @devkapilbansal 🤗 I just want to remind you of when this is done to put README instructions of how to run this project with docker 😉

**Comment by devkapilbansal** [comment]
@isabelcosta @vj-codes this is ready for review

**Review by codesankalp — DISMISSED** [review]
image
image
Tested and Working fine @devkapilbansal :smile: .

File: README.md [hunk 1]
```diff
@@ -52,6 +52,25 @@ Next follow these instructions.
...
+- Go to ``main/settings.py``
+- Under databases change host to db.
+
+3. **Build docker image**
+
+Execute ``sudo docker-compose run web`` to build the image.
+
+4. **Run docker image**
+
+Run ``sudo docker-compose up`` from the root directory of project.
```
Comment (isabelcosta) [inline README.md]: as step 4 you could have an additional step with the link where the server is running, so we don't have to look at dockerfiles to test it live. is it localhost:8000?

File: README.md [hunk 2]
```diff
@@ -52,6 +52,25 @@ Next follow these instructions.
 5. To run the migrations run: `python manage.py migrate`
 6. You can terminate the process by `Ctrl+C` in your terminal.
 
+### Run with Docker
+1. Make sure latest version of docker-compose is installed by running ``docker-compose -v``. If it is not,install using ``pip install -U docker-compose``
```
Comment (isabelcosta) [inline README.md]:
```suggestion
1. Make sure the latest version of docker-compose is installed by running ``docker-compose -v``. If it is not, install using ``pip install -U docker-compose``
```

File: README.md [hunk 3]
```diff
@@ -52,6 +52,25 @@ Next follow these instructions.
 5. To run the migrations run: `python manage.py migrate`
 6. You can terminate the process by `Ctrl+C` in your terminal.
 
+### Run with Docker
```
Comment (isabelcosta) [inline README.md:81]:
```suggestion
### Run with Docker

```

**Review by isabelcosta — CHANGES_REQUESTED** [review]
Tiny changes, I would love to test this out :)
Thank you for doing this @devkapilbansal !

File: README.md (hunk 1 above)
Comment (devkapilbansal) [reply README.md]: Yes it is localhost:8000

**Comment by devkapilbansal** [comment]
> Tiny changes, I would love to test this out :)
> Thank you for doing this @devkapilbansal !

Thanks for the review, will make the changes soon.

**Comment by devkapilbansal** [comment]
@isabelcosta please review it again

**Comment by isabelcosta** [comment]
@codesankalp can you please add a review again :)

File: README.md [hunk 4]
```diff
@@ -52,6 +52,26 @@ Next follow these instructions.
...
+1. Make sure the latest version of docker-compose is installed by running ``docker-compose -v``. If it is not, install using ``pip install -U docker-compose``
+
+2. Change Database HOST in ``settings.py``
+
+- Go to ``main/settings.py``
+- Under databases change host to db.
+
+3. **Build docker image**
+
+Execute ``sudo docker-compose run web`` to build the image.
```
Comment (codesankalp) [inline README.md:66]: @devkapilbansal This command is not for building the docker image.
```suggestion
Execute ``sudo docker-compose up`` to build the image.
```

File: README.md [hunk 5]
```diff
@@ -52,6 +52,26 @@ Next follow these instructions.
...
+
+4. **Run docker image**
+
+Run ``sudo docker-compose up`` from the root directory of project. Navigate to `http://localhost:8000` in your browser.
```
Comment (codesankalp) [inline README.md:70]: @devkapilbansal This should be used for running server.
```suggestion
Run ``sudo docker-compose run websudo ss -lptn 'sport = :5432'`` from the root directory of project. Navigate to `http://localhost:8000` in your browser.
```

File: README.md [hunk 6]
```diff
@@ -52,6 +52,26 @@ Next follow these instructions.
...
+
```
Comment (codesankalp) [inline README.md:100]: There should be a command for migrating in docker. Just a suggestion :)
```suggestion

5. **Migrating Database**

To run the migrations run ``sudo docker-compose run web python manage.py migrate``.
```

**Review by codesankalp — CHANGES_REQUESTED** [review]
It is working fine @devkapilbansal :+1: 
image
image
Testing and Review have done @isabelcosta . :smile:

**Review by codesankalp — DISMISSED** [review]

**Comment by isabelcosta** [comment]
I tried this out and it exploded. Do I need environment variables for this to work? @codesankalp @devkapilbansal I don't find documentation about it, so we should have an issue for this

**Comment by devkapilbansal** [comment]
> I tried this out and it exploded. Do I need environment variables for this to work? @codesankalp @devkapilbansal I don't find documentation about it, so we should have an issue for this

@isabelcosta No, with docker you may not need to setup anything. Can you please explain the exact error that occurs? I or someone else may be able to look into it then

File: README.md (hunk 6 above)
Comment (devkapilbansal) [reply README.md:100]: @codesankalp the user can log into the container once build and run any command he wants.

`docker exec -it {container_name} /bin/bash` will open a terminal for the same.

**Comment by isabelcosta** [comment]
> > I tried this out and it exploded. Do I need environment variables for this to work? @codesankalp @devkapilbansal I don't find documentation about it, so we should have an issue for this
> 
> @isabelcosta No, with docker you may not need to setup anything. Can you please explain the exact error that occurs? I or someone else may be able to look into it then

image

I ran `docker-compose up --build`

and then http://localhost:8000/ leads no where

**Comment by devkapilbansal** [comment]
@isabelcosta can you please confirm that Database host is set to **`db`** instead of **`localhost`** in `main/settings.py` similar to the below configuration

database

**Comment by codesankalp** [comment]
Yes @isabelcosta , when I ran `docker-compose up --build` without changing host in settings.py i.e from `localhost` to `db` it shows me the same output as yours.

image

So changing it to `db` will solve the problem.

**Comment by isabelcosta** [comment]
OK, I did forget about that step! thank you @codesankalp and @devkapilbansal !
Now I have another issue, do I have to setup an API key? for example from zulip, I have seen doubts about this before.

image

**Comment by devkapilbansal** [comment]
Yes, @isabelcosta you need to go to user settings in zulip and request the API key. Then download the api key and keep the folder in root directory. There was discussion about user id and some other stuffs too on zulip about which I am still unsure. 😕

**Comment by codesankalp** [comment]
I think we should provide a bot test API key (of anitab-org-zulip) in the code so that new users will not face any problems regarding downloading of API key.
What's your thought @isabelcosta and @devkapilbansal ?

**Comment by isabelcosta** [comment]
I haven't yet understood how to get the variables, meanwhile, if you get someone else from the community that can confirm this is working and approves, I am more than happy to merge cc @devkapilbansal @codesankalp

File: README.md [hunk 7]
```diff
@@ -52,6 +52,26 @@ Next follow these instructions.
 5. To run the migrations run: `python manage.py migrate`
 6. You can terminate the process by `Ctrl+C` in your terminal.
 
+### Run with Docker
+
+1. Make sure the latest version of docker-compose is installed by running ``docker-compose -v``. If it is not, install using ``pip install -U docker-compose``
```
Comment (ashwinijha6) [inline README.md]: @devkapilbansal just add `sudo apt install docker.io` for installing docker before the 1st step in the README for **Run with Docker**. Also u can include the version check `docker -v`  as u have mentioned in the step above.

**Review by ashwinijha6 — CHANGES_REQUESTED** [review]

**Comment by ashwinijha6** [comment]
Also @devkapilbansal  just check once with the command `pip install -U docker-compose` though there is no problem with the command but after installing using this command the build shows unknown while checking for the version.
I installed docker-compose using https://docs.docker.com/compose/install/

**Review by jalajcodes — DISMISSED** [review]
I finally got on a stable connection and I was able to run it correctly. Nice work @devkapilbansal 🎉

image

image

**Comment by jalajcodes** [comment]
Also, are there any plans to containerize the frontend? I'd love to work on that

File: README.md [hunk 8]
```diff
@@ -60,7 +60,7 @@ Next follow these instructions.
     python manage.py runserver
     ```
 8. Navigate to `http://localhost:8000/` in your browser.
-9. To change the port you may run `python manage.py runserver 0.0.0.0:<port_number>`
```
Comment (devkapilbansal) [inline README.md:63]: Removed `0.0.0.0` as it is not required

File: README.md [hunk 9]
```diff
@@ -78,6 +78,30 @@ Follow the given instructions for Login into the app.
...
+
+Execute ``sudo docker-compose run web`` to build the image.
+
+4. **Run docker image**
+
+Run ``sudo docker-compose up`` from the root directory of project. Navigate to `http://localhost:8000` in your browser.
+
+**Note** 
+- Run `docker-compose up --build` to rebuild the image.
+- To **interact** with docker containers use ``docker exec -it {container id} bash`` from your terminal. Container id can be found using ``docker ps``.
```
Comment (devkapilbansal) [inline README.md:103]: @codesankalp Added this command if someone wants to run commands like `python manage.py migrate`.

**Review by devkapilbansal — COMMENTED** [review]
@isabelcosta @codesankalp @ashwinijha6 @jalajcodes can you please review this pull request again. I rebase the branch and improved the instructions a bit more.

**Review by ashwinijha6 — DISMISSED** [review]
Looks good to me @devkapilbansal 👍

File: README.md (hunk 9 above)
Comment (codesankalp) [reply README.md:103]: Thanks @devkapilbansal , for adding this.

**Review by codesankalp — DISMISSED** [review]
I tested it again and it is working fine.
LGTM @devkapilbansal  and @isabelcosta .

**Comment by isabelcosta** [comment]
I just fixed the little merge conflict on README, will merge since everyone had approved right before the other merge 🙈

**Comment by isabelcosta** [comment]
Awesome to see this project with a dockerfile 💪🏾 
Thank you for working on it @devkapilbansal !

**Pull request merged**